src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -491,6 +542,124 @@ private int FindEndPosition(ReadOnlySpan<char> input, int i)
             return false;
         }
 
+        /// <summary>Find match end position using the original pattern, end position is known to exist. This version also produces captures.</summary>
+        /// <param name="input">input span</param>
+        /// <param name="i">inclusive start position</param>
+        /// <param name="resultRegisters">out parameter for the final register values, which indicate capture starts and ends</param>
+        /// <returns></returns>
+        private int FindEndPositionCapturing(ReadOnlySpan<char> input, int i, out Registers resultRegisters)
+        {
+            int i_end = input.Length;
+            Registers endRegisters = default(Registers);
+
+            // Pick the correct start state based on previous character kind.
+            uint prevCharKind = GetCharKind(input, i - 1);
+            DfaMatchingState<TSetType> state = _initialStates[prevCharKind];
+            // Initialize registers with -1, which means "not seen yet"
+            Registers initialRegisters = new Registers
+            {
+                captureStarts = new int[_capsize],
+                captureEnds = new int[_capsize],
+            };
+            Array.Fill(initialRegisters.captureStarts, -1);
+            Array.Fill(initialRegisters.captureEnds, -1);
+
+            if (state.IsNullable(GetCharKind(input, i)))
+            {
+                // Empty match exists because the initial state is accepting.
+                i_end = i - 1;
+                // Some empty groups may still be captured, apply effects as necessary
+                endRegisters = initialRegisters.Clone();
+                state.Node.ApplyEffects(effect => endRegisters.ApplyEffect(effect, i), CharKind.Context(state.PrevCharKind, GetCharKind(input, i)));
+            }
+
+            // Use two maps from state IDs to register values for the current and next set of states.
+            // Note that these maps use insertion order, which is used to maintain priorities between states in a way
+            // that matches the order the backtracking engines visit paths.
+            SparseIntMap<Registers> current = new(), next = new();
+            current.Add(state.Id, initialRegisters);
+
+            while (i < input.Length)
```

Comment:
```suggestion
            while ((uint)i < (uint)input.Length)
```
Just helps avoid subsequent bounds checks in `input[i]`